### ISSUE TITLE
[feature request] Add support for a custom DatasetFetcher in DataLoader 

### ISSUE THREAD

**Issue by Abermal** [pytorch/pytorch#80134]
### 🚀 The feature, motivation and pitch

I'm working on a time-series dataset, and I stumbled upon a limitation withing pytorch's DataLoader that I'd like to address here.

I have a custom map-style dataset that implements a `__getitem__` and `__len__` methods.

`__getitem__`  has a following signature and returns by default features of shape `(seq_length, n_features)` and an integer label.
```python
def __getitem__(self, i: int, seq_length: int = 100):
    # ....
    return features, label
```
Sequence length is fixed by default, but can be changed by passing a `seq_length` argument.

I'm using an LSTM + Transformer model that is able to process an input of variable sequence length.\
BUT I'm not leveraging this feature of my model because pytorch's DataLoader expects all the samples returned by the dataset to be of the same shape.
A traditional approach to solving this problem is **sequence padding**. This doesn't really solve the problem in my opinion (the model still gets constant sequence length as an input) and leads to a range of further questions (which values to use for padding, how to interpolate ect.).

**What I would like to do instead** is to use a custom DatasetFetcher to create batches that have constant sequence length within a batch, but **varying** sequence length over the dataset.\
So let's say or batch has a shape `batch_size, seq_length, n_features`.  (Let batch_size=32, n_features=10)\
Our DataLoader will sample indices for the first batch and then fetch samples with a constant `seq_length=100`, which will result in a batch of shape `32, 100, 10`.\
On the next batch, it will randomly (or by any other rule) pick a new sequence length (let's say 80) and create a batch of shape `32, 80, 10`.

```python
# Default behaviour
data = [self.dataset[idx] for idx in possibly_batched_index]
# New behaviour
data = [self.dataset.__getitem__(idx, seq_length) for idx in possibly_batched_index]
```

I'm pretty sure it will have a regularization effect.

**Comment by NivekT** [comment]
Have you considered pass in a custom `collate_fn` to `DataLoader` that purpose? The input to `collate_fn` is a batch; you are free to pick a sequence length and pad/trim the sequence in there as you need. You can also import and call the `torch.utils.data.default_collate` after your custom step if you'd like as well.
